src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,585 @@
+        remote_readbuffer = nullptr;
+        lock.unlock();
+        more_data_signal.notify_all();
+        finish_callback(this);
+        LOG_TRACE(
+            &Poco::Logger::get("RemoteCacheController"),
+            "finish download.{} into {}. size:{} ",
+            remote_path, local_path.string(), current_offset);
+    };
+    RemoteReadBufferCache::instance().GetThreadPool()->scheduleOrThrow(task);
+}
+
+void RemoteCacheController::flush(bool need_flush_meta_)
+{
+    if (out_file)
+    {
+        out_file->flush();
+    }
+
+    if (!need_flush_meta_)
+        return;
+    Poco::JSON::Object jobj;
+    jobj.set("schema", schema);
+    jobj.set("cluster", cluster);
+    jobj.set("remote_path", remote_path);
+    jobj.set("downloaded", download_finished ? "true" : "false");
+    jobj.set("last_modification_timestamp", last_modification_timestamp);
+    std::stringstream buf;// STYLE_CHECK_ALLOW_STD_STRING_STREAM
+    jobj.stringify(buf);
+
+    std::ofstream meta_file(local_path / "meta.txt", std::ios::out);
+    meta_file << buf.str();
+    meta_file.close();
+}
+
+RemoteCacheController::~RemoteCacheController() = default;
+void RemoteCacheController::close()
+{
+    // delete the directory
+    LOG_TRACE(&Poco::Logger::get("RemoteCacheController"), "release local resource: " + remote_path + ", " + local_path.string());
+    std::filesystem::remove_all(local_path);
+}
+
+std::tuple<FILE *, std::filesystem::path> RemoteCacheController::allocFile()
+{
+    std::filesystem::path result_local_path;
+    if (download_finished)
+        result_local_path = local_path /  "data.bin";
+    FILE * fs = fopen((local_path / "data.bin").string().c_str(), "r");
+    if (fs == nullptr)
+        return {fs, result_local_path};
+    std::lock_guard lock{mutex};
+    opened_file_streams.insert(fs);
+    return {fs, result_local_path};
+}
+
+void RemoteCacheController::deallocFile(FILE * fs)
+{
+    {
+        std::lock_guard lock{mutex};
+        auto it = opened_file_streams.find(fs);
+        if (it == opened_file_streams.end())
+        {
+            std::string err = "try to close an invalid file " + remote_path;
+            throw Exception(err, ErrorCodes::BAD_ARGUMENTS);
+        }
+        opened_file_streams.erase(it);
+    }
+    fclose(fs);
+}
+
+LocalCachedFileReader::LocalCachedFileReader(RemoteCacheController * cntrl_, size_t size_)
+    : offset(0), file_size(size_), fs(nullptr), controller(cntrl_)
+{
+    std::tie(fs, local_path) = controller->allocFile();
+    if (fs == nullptr)
+        throw Exception("alloc file failed.", ErrorCodes::BAD_GET);
+}
+LocalCachedFileReader::~LocalCachedFileReader()
+{
+    controller->deallocFile(fs);
+}
+
+size_t LocalCachedFileReader::read(char * buf, size_t size)
+{
+    auto wret = controller->waitMoreData(offset, offset + size);
+    if (wret != RemoteReadBufferCacheError::OK)
+        return 0;
+    std::lock_guard lock(mutex);
+    auto ret_size = fread(buf, 1, size, fs);
```

Comment:
As I have explained before, it's not easy to make the memory in ReadBufferFromFileBase be shared outside, and we have to copy data from one to anther . 
If you think this overhead is OK, I will rewrite it.

Reply:
>  it's not easy to make the memory in ReadBufferFromFileBase be shared outside

where do you need it to be shared?

Reply:
I mean that make RemoteReadBuffer and ReadBufferFromFileBase use the same memory space

Reply:
why not? can you just use `set` method of ReadBuffer?

Reply:
I check it again. ReadBufferFromFileBase is seekable, it may work if the cached file is all downloaded.
But in our design, we don't need the file be all downloaded before the reader can read it which make the process not be blocked too long time.